src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregatorStore.cs

```diff
@@ -0,0 +1,472 @@
+        {
+            object? stateUnion = _stateUnion;
+            if (stateUnion is TAggregator agg)
+            {
+                IAggregationStatistics stats = agg.Collect();
+                visitFunc(new LabeledAggregationStatistics(stats));
+            }
+            else if (stateUnion is FixedSizeLabelNameDictionary<StringSequence1, ObjectSequence1, TAggregator> aggs1)
+            {
+                aggs1.Collect(visitFunc);
+            }
+            else if (stateUnion is FixedSizeLabelNameDictionary<StringSequence2, ObjectSequence2, TAggregator> aggs2)
+            {
+                aggs2.Collect(visitFunc);
+            }
+            else if (stateUnion is FixedSizeLabelNameDictionary<StringSequence3, ObjectSequence3, TAggregator> aggs3)
+            {
+                aggs3.Collect(visitFunc);
+            }
+            else if (stateUnion is FixedSizeLabelNameDictionary<StringSequenceMany, ObjectSequenceMany, TAggregator> aggsMany)
+            {
+                aggsMany.Collect(visitFunc);
+            }
+            else if (stateUnion is MultiSizeLabelNameDictionary<TAggregator> aggsMultiSize)
+            {
+                aggsMultiSize.Collect(visitFunc);
+            }
+        }
+
+
+        public TAggregator GetAggregator()
+        {
+            while (true)
+            {
+                object? state = _stateUnion;
+                if (state == null)
+                {
+                    TAggregator newState = _createAggregatorFunc();
+                    Interlocked.CompareExchange(ref _stateUnion, newState, null);
+                    continue;
+                }
+                else if (state is TAggregator aggState)
+                {
+                    return aggState;
+                }
+                else if (state is MultiSizeLabelNameDictionary<TAggregator> multiSizeState)
+                {
+                    return multiSizeState.GetNoLabelAggregator(_createAggregatorFunc);
+                }
+                else
+                {
+                    MultiSizeLabelNameDictionary<TAggregator> newState = new(state);
+                    Interlocked.CompareExchange(ref _stateUnion, newState, state);
+                    continue;
+                }
+            }
+        }
+
+        public ConcurrentDictionary<TObjectSequence, TAggregator> GetLabelValuesDictionary<TStringSequence, TObjectSequence>(in TStringSequence names)
+            where TStringSequence : IStringSequence, IEquatable<TStringSequence>
+            where TObjectSequence : IObjectSequence, IEquatable<TObjectSequence>
+        {
+            while (true)
+            {
+                object? state = _stateUnion;
+                if (state == null)
+                {
+                    FixedSizeLabelNameDictionary<TStringSequence, TObjectSequence, TAggregator> newState = new();
+                    Interlocked.CompareExchange(ref _stateUnion, newState, null);
```

Comment:
Same suggestions here.  Races should be rare (yes?) so we can optimize for the success case and avoid the extra loop around, type checks, and memory barrier (on arm).